### ISSUE TITLE
ARIA role Attribute

### ISSUE THREAD

**Issue by ways2read** [ways2read/InDesignA11Y#31]
Knowledge base reference: http://kb.daisy.org/publishing/docs/html/roles.html

**Comment by gregoriopellegrino** [comment]
InDesign currently uses `epub:type` attributes to provide semantic information in the EPUB export about document content. While `epub:type` was intended to define the structure of digital publications by conveying structural meaning, it is now considered deprecated. Modern accessibility standards rely on ARIA (Accessible Rich Internet Applications) roles and attributes, which provide better support for assistive technologies and are widely adopted across digital platforms.

## EPUB:TYPE IMPLEMENTATION
The `epub:type` attribute can be assigned via Object Export Options at either the object or text frame level.

![Image](https://github.com/user-attachments/assets/941fa095-0be9-4d9f-a6b2-ce285bc5beb0)

However, this implementation has several limitations:

1.  The `epub:type` attribute is considered deprecated and is no supported by assistive technologies
2.  There is no clear role management when content is distributed across multiple threaded text frames
3.  Lacks the ability to assign semantic roles to different levels of content (paragraph, character) in a granular way

## WHY ARIA ROLES ARE IMPORTANT
ARIA roles represent the current standard for making digital documents
accessible. Their importance stems from several factors:

-   They allow screen readers and assistive technologies to understand the semantic structure of the document
-   They provide a standardized way to identify different types of  content (notes, citations, glossaries, etc.)
-   They support efficient navigation through landmarks
-   They are compatible with modern web standards and widely supported

## WHAT SHOULD BE IMPLEMENTED

1.  Support for ARIA Roles
    -   In addition to defining `epub:type` the user should be able to apply ARIA roles
2.  Multi-level support for ARIA Roles
    -   At paragraph level: integration into paragraph styles to assign roles to entire blocks of text; it may be added to the Export Tagging tab:<br>![Image](https://github.com/user-attachments/assets/ed7630c9-7335-40a2-884c-45e1c2d71d78)
    -   At the character level: ability to assign roles to inline elements such as glossary term references or citations; it may be added to the Export Tagging tab as well;
    -   At object/frame level: maintain and enhance existing support for objects (in the Object Export Options window)
3.  Intelligent Text Frame Management.
    -   Allow specifying how roles should be handled in threads (inherited from the first frame in the thread or individual)
    -   Manage export correctly when frames are split into separate HTML files

## ARIA ROLES TO PROVIDE

Below is a list of ARIA roles to be provided to content creators to assign to objects, paragraphs, and characters:

-   [doc-abstract](https://www.w3.org/TR/dpub-aria-1.1/#doc-abstract)
-   [doc-acknowledgments](https://www.w3.org/TR/dpub-aria-1.1/#doc-acknowledgments)
-   [doc-afterword](https://www.w3.org/TR/dpub-aria-1.1/#doc-afterword)
-   [doc-appendix](https://www.w3.org/TR/dpub-aria-1.1/#doc-appendix)
-   [doc-backlink](https://www.w3.org/TR/dpub-aria-1.1/#doc-backlink)
-   [doc-bibliography](https://www.w3.org/TR/dpub-aria-1.1/#doc-bibliography)
-   [doc-biblioref](https://www.w3.org/TR/dpub-aria-1.1/#doc-biblioref)
-   [doc-chapter](https://www.w3.org/TR/dpub-aria-1.1/#doc-chapter)
-   [doc-colophon](https://www.w3.org/TR/dpub-aria-1.1/#doc-colophon)
-   [doc-conclusion](https://www.w3.org/TR/dpub-aria-1.1/#doc-conclusion)
-   [doc-credit](https://www.w3.org/TR/dpub-aria-1.1/#doc-credit)
-   [doc-credits](https://www.w3.org/TR/dpub-aria-1.1/#doc-credits)
-   [doc-dedication](https://www.w3.org/TR/dpub-aria-1.1/#doc-dedication)
-   [doc-epigraph](https://www.w3.org/TR/dpub-aria-1.1/#doc-epigraph)
-   [doc-epilogue](https://www.w3.org/TR/dpub-aria-1.1/#doc-epilogue)
-   [doc-errata](https://www.w3.org/TR/dpub-aria-1.1/#doc-errata)
-   [doc-example](https://www.w3.org/TR/dpub-aria-1.1/#doc-example)
-   [doc-foreword](https://www.w3.org/TR/dpub-aria-1.1/#doc-foreword)
-   [doc-glossary](https://www.w3.org/TR/dpub-aria-1.1/#doc-glossary)
-   [doc-glossref](https://www.w3.org/TR/dpub-aria-1.1/#doc-glossref)
-   [doc-index](https://www.w3.org/TR/dpub-aria-1.1/#doc-index)
-   [doc-introduction](https://www.w3.org/TR/dpub-aria-1.1/#doc-introduction)
-   [doc-notice](https://www.w3.org/TR/dpub-aria-1.1/#doc-notice)
-   [doc-pagebreak](https://www.w3.org/TR/dpub-aria-1.1/#doc-pagebreak)
-   [doc-pagefooter](https://www.w3.org/TR/dpub-aria-1.1/#doc-pagefooter)
-   [doc-pageheader](https://www.w3.org/TR/dpub-aria-1.1/#doc-pageheader)
-   [doc-part](https://www.w3.org/TR/dpub-aria-1.1/#doc-part)
-   [doc-preface](https://www.w3.org/TR/dpub-aria-1.1/#doc-preface)
-   [doc-prologue](https://www.w3.org/TR/dpub-aria-1.1/#doc-prologue)
-   [doc-pullquote](https://www.w3.org/TR/dpub-aria-1.1/#doc-pullquote)
-   [doc-qna](https://www.w3.org/TR/dpub-aria-1.1/#doc-qna)
-   [doc-subtitle](https://www.w3.org/TR/dpub-aria-1.1/#doc-subtitle)
-   [doc-tip](https://www.w3.org/TR/dpub-aria-1.1/#doc-tip)
-   [math](https://www.w3.org/TR/wai-aria/#math)
-   [note](https://www.w3.org/TR/wai-aria/#note)
-   [presentation](https://www.w3.org/TR/wai-aria/#presentation)
-   [region](https://www.w3.org/TR/wai-aria/#region)
-   [separator](https://www.w3.org/TR/wai-aria/#separator)

**Comment by Nainabhandari24** [comment]
Hi @gregoriopellegrino, @LauraB7 @ways2read @jonaslil, 

We are pleased to share that the "ARIA role feature" is now available in the pre-release build (ID 21.0.0.146).

Thanks,
Naina

**Comment by Nainabhandari24** [comment]
Hi @LauraB7,
Could you please help us by providing name of any screen reader which is helpful to read "DAISY / EPUB ARIA document roles"?

Thanks,
Naina

**Comment by LauraB7** [comment]
The Daisy testing group had good results with Talkback on Android using Chrome webviews. See results in [this spreadsheet](https://github.com/daisy/transitiontoepub/blob/main/dpub-aria-testing/test-results/DPUB-ARIA-test-results.csv).

**Comment by Nainabhandari24** [comment]
Hi @LauraB7, 

Thank you for sharing the test results. It appears that certain roles are not consistently announced by screen readers such as JAWS and NVDA. Could you please confirm whether any fixes are required from our side? If so, we would appreciate your guidance on how these can be implemented.

Thanks,
Naina

**Comment by gregoriopellegrino** [comment]
Dear @Nainabhandari24,
thank you for implementing the ARIA role feature in the pre-release build. This is a significant step forward for digital accessibility in InDesign. We'd like to provide some feedback and suggestions based on our testing and experience with ARIA roles in EPUBs.

## Assistive Technology Support Context
Regarding the screen reader support concerns mentioned, it's important to understand that DPUB ARIA support by assistive technologies is still patchy, largely because content with DPUB ARIA roles is still scarce. Very few authoring tools can produce them currently. With InDesign's development in this area, we expect a significant push in support from assistive technologies. The DAISY Consortium continuously works to raise awareness among assistive technology developers about better supporting for DPUB ARIA roles, and having a major tool like InDesign supporting these roles will accelerate adoption.

## Suggested Role Organization

Working with the roles, we've realized that not all roles are appropriate for all elements. We'd like to suggest dividing the suggestion lists as follows:

### For Paragraph Style

- doc-abstract
- doc-acknowledgments
- doc-credit
- doc-dedication
- doc-epigraph
- doc-errata
- doc-example
- doc-notice
- doc-pagebreak
- doc-pullquote
- doc-subtitle
- doc-tip
- math
- note
- presentation

### For Character Style

- doc-pagebreak
- doc-tip
- math
- note
- presentation

### For Objects (Object Export Options, Object Style)

- doc-abstract
- doc-acknowledgments
- doc-afterword
- doc-appendix
- doc-bibliography
- doc-chapter
- doc-colophon
- doc-conclusion
- doc-credits
- doc-dedication
- doc-epigraph
- doc-epilogue
- doc-errata
- doc-example
- doc-foreword
- doc-glossary
- doc-index
- doc-introduction
- doc-notice
- doc-pagefooter
- doc-pageheader
- doc-part
- doc-preface
- doc-prologue
- doc-pullquote
- doc-qna
- doc-tip
- doc-toc
- math
- note
- presentation
- region
- separator

### For Links (Hyperlinks and Cross-References)

- doc-backlink
- doc-biblioref
- doc-glossref

## Hyperlinks and Cross-References Support

We would need the ability to add roles to hyperlinks and cross references. Both interfaces already have an "Accessibility" tab where this could be added.

<img width="679" height="566" alt="screenshot of New Hyperlink window" src="https://github.com/user-attachments/assets/e04c5925-e29c-4514-8eec-1c1755edb39a" />

## Threaded Text Frame Issue

We're experiencing a problem with threaded text frames where text flows throughout the book (a very common case) and XHTML files within the EPUB are split based on "Split Document (EPUB only)".

In our test case ([test-aria-role-threaded-frames.zip](https://github.com/user-attachments/files/22613877/test-aria-role-threaded-frames.zip)):

- The first frame has the `doc-chapter` role and subsequent frames inherit it
- The last frame is assigned `doc-conclusion` but this role is not being correctly exported (we get `doc-chapter`)

Thank you for your continued work on this important feature!

Gregorio

**Comment by Nainabhandari24** [comment]
Thanks for the feedback — we’re looking into it.



Naina

**Comment by Nainabhandari24** [comment]
Hi All,

We are pleased to inform, we have implemented ARIA role and ARIA label in hyperlinks and cross-references in the ID21.1.0.42 pre-release build.
Please try the feature and share your feedback.

Thanks,
Naina

**Comment by Nainabhandari24** [comment]
Hi Team,

I hope you have validated "ARIA role Attribute" ticket. We request you to close the ticket.


Thanks,
Naina

**Comment by gregoriopellegrino** [comment]
Dear @Nainabhandari24,
Thank you for the continued development on ARIA role support. We've tested the latest pre-release build and here's our feedback on the requested features:

Implemented Successfully:

- Reorganize roles based on style type (object, paragraph, character, and links)
- ARIA roles for hyperlinks and cross-references

 
Still Outstanding:

- Threaded Text Frame Issue - as mentioned in our [previous feedback](https://github.com/ways2read/InDesignA11Y/issues/31#issuecomment-3350562402) (September 30), when text flows through threaded frames and documents are split based on "Split Document (EPUB only)", the ARIA role assigned to the last frame is not being correctly exported. Instead of receiving the assigned role (e.g., doc-conclusion), it inherits the role from the first frame (e.g., doc-chapter).

 
This is a critical issue for books and long documents where content flows through multiple frames, which is a very common use case in InDesign.

Could you please provide an update on when this threaded frame issue might be addressed?

Thanks again for your work on this important accessibility feature!

**Comment by raman211** [comment]
Thank you @gregoriopellegrino for the update.

**Still Outstanding:**
Threaded Text Frame Issue support is currently at the Object Export Options Dialog level in the pre-release builds of ID21.2.18 
The support at the Object Style is part of our internal builds, which will be part of ID21.2

CC: @ways2read @LauraB7 @jonaslil @Nainabhandari24

**Comment by raman211** [comment]
@gregoriopellegrino 
**Still Outstanding:** The fix for Threaded Text Frame Issue is already part of ID21.2.0.27 pre-release builds.

Please help us by sharing your feedback on this.
CC: @ways2read @LauraB7 @jonaslil @Nainabhandari24

**Comment by Nainabhandari24** [comment]
Hi @gregoriopellegrino @ways2read ,

We are pleased to announce that InDesign 21.2 is now available worldwide.
We look forward to your feedback.

CC: @raman211

Thanks,
Naina

**Comment by gregoriopellegrino** [comment]
Thank you, we have tested it and it works well for us. Great job!
